### PR TITLE
Fail2ban with HTTP status and logger (#2291 #2384)

### PR DESCRIPTION
**Pull request by yweber**
I'm not sure if this patch is acceptable. I never wrote golang before but I was needing this features.
- introduce a new context function RenderWithErrStatus that allows to
  specify an HTTP status when rendering errors on form datas
  - status 403 set on repository permission denied
  - status 401 set on authentication failure
  - status 401 set when no associated email found on forgotten password
    procedure
- add warning log message, containing client IP address, for
  authentication failure
- add warning log message in routers/user/auth.go AutoSignIn function
- add http_status int in returned values for routers/user/auth.go
  AutoSignIn function
- change default log level to Info

### CODE REVIEW

File: conf/app.ini [hunk 1]
```diff
@@ -257,7 +257,7 @@ MODE = console
 ; Buffer length of channel, keep it as it is if you don't know what it is.
 BUFFER_LEN = 10000
 ; Either "Trace", "Debug", "Info", "Warn", "Error", "Critical", default is "Trace"
-LEVEL = Trace
+LEVEL = Info
```
Comment (Unknwon) [inline conf/app.ini]: Why change this line?

Comment (yweber) [inline conf/app.ini]: I was thinking that was a better default log level. But at this time I wasn't knowing I will push on develop branch, where Trace is indeed a relevant log level.  Do you want me to delete this change ?

Comment (Unknwon) [inline conf/app.ini]: Obviously you would have to delete this change..

**Comment by rugk** [comment]
:+1: for this.
@yweber Do you also have a Fail2Ban RegExp for this?
If so you might also want to add it to the doc somewhere.

**Comment by yweber** [comment]
@rugk Thx ! And on my installation I don't use fail2ban directly on gogs logfiles but on the nginx reverse-proxy access log ( I filter on HTTP status ). I have written and quickly tested this RegExp for gogs logfiles : `^[^\[]*\[W\] <HOST> Authentication failure for user .*$`

Writing this RegExp I figured out that on password retrieval failure (when password retrieval is asked for a non registered email address) no log message is produced and that a different HTTP status is returned (401 instead of 200). This behavior can allow an attacker to find if an email is registered or not.
What do you think ? Is it a better idea to invert this behavior, returning HTTP 200 and producing a log message containing client IP when password retrieval fails ?

**Comment by rugk** [comment]
> Is it a better idea to invert this behavior, returning HTTP 200 and producing a log message containing client IP when password retrieval fails ?

I think in both cases either 401 or 403 should be returned and you can additionally write the log file.

**Comment by Unknwon** [comment]
Please sign the CLA.

**Comment by AndreaGiardini** [comment]
Any update on this?

**Comment by dtrv** [comment]
+1

**Comment by rugk** [comment]
Maybe @yweber could try to introduce this PR in the fork @gitea.

**Comment by Fastidious** [comment]
I find it of poor taste, @rugk, to come to a project, and try to send people to another. You could have opened an issue on Gitea about this, and mention the OP in it.

**Comment by rugk** [comment]
It does not really make a difference. Also it's @yweber' choice…

**Comment by yweber** [comment]
The merge commit https://github.com/gogits/gogs/pull/3357/commits/47fc72f3d1422426866d430d806ac44d50081cdd is just about solving merge conflicts and making the compilation (and apparently the software) work :
I didn't take the time to check if some other 4xx, 5xx HTTP status or other log lines should be emitted with code changes. I also did not check if a function was implemented achieving the same goal than
[RenderWithErrStatus]( https://github.com/gogits/gogs/pull/3357/commits/47fc72f3d1422426866d430d806ac44d50081cdd#diff-94ab5308b984ba8eb1cb1aa88d0487d7R104 ).

**Comment by fizzyade** [comment]
Why hasn't this made it into the source?  This is a useful feature.

**Pull request closed without merging**